video/agon_audio.h

```diff
@@ -1,59 +1,363 @@
+		return false;
+	}
+	return (elapsed >= this->_duration);
+}
+
+bool loopy = false;
+
 void audio_channel::loop() {
-	if(this->_flag > 0) {
-		debug_log("audio_driver: play %d,%d,%d,%d\n\r", this->_channel, this->_volume, this->_frequency, this->_duration);			
-		this->_waveform->setVolume(this->_volume);
-		this->_waveform->setFrequency(this->_frequency);
-		this->_waveform->enable(true);
-		vTaskDelay(this->_duration);
-		this->_waveform->enable(false);
-		debug_log("audio_driver: end\n\r");			
-		this->_flag = 0; 
+	if (!loopy) {
+		loopy = true;
+		debug_log("audio_driver: loop %d\n\r", this->_channel);
+	}
+	switch (this->_state) {
+		case AUDIO_STATE_PENDING:
+			debug_log("audio_driver: play %d,%d,%d,%d\n\r", this->_channel, this->_volume, this->_frequency, this->_duration);
+			// we have a new note to play
+			this->_startTime = millis();
+			// set our initial volume and frequency
+			this->_waveform->setVolume(this->getVolume(0));
+			this->_waveform->setFrequency(this->getFrequency(0));
+			this->_waveform->enable(true);
+			// if we have an envelope then we loop, otherwise just delay for duration			
+			if (this->_volumeEnvelope) {
+				this->_state = AUDIO_STATE_PLAY_LOOP;
+			} else {
+				this->_state = AUDIO_STATE_PLAYING;
+				// if delay value is negative then this delays for a super long time
+				vTaskDelay(pdMS_TO_TICKS(this->_duration));
+			}
+			break;
+		case AUDIO_STATE_PLAYING:
+			if (this->_duration >= 0) {
+				// simple playback - delay until we have reached our duration
+				word elapsed = millis() - this->_startTime;
+				debug_log("audio_driver: elapsed %d\n\r", elapsed);
+				if (elapsed >= this->_duration) {
+					this->_waveform->enable(false);
+					debug_log("audio_driver: end\n\r");
+					this->_state = AUDIO_STATE_IDLE;
+				} else {
+					debug_log("audio_driver: loop (%d)\n\r", this->_duration - elapsed);
+					vTaskDelay(pdMS_TO_TICKS(this->_duration - elapsed));
+				}
+			} else {
+				// our duration is indefinite, so delay for a long time
+				debug_log("audio_driver: loop (indefinite playback)\n\r");
+				vTaskDelay(pdMS_TO_TICKS(-1));
+			}
+			break;
+		// loop and release states used for envelopes
+		case AUDIO_STATE_PLAY_LOOP: {
+			word elapsed = millis() - this->_startTime;
+			if (isReleasing(elapsed)) {
+				debug_log("audio_driver: releasing...\n\r");
+				this->_state = AUDIO_STATE_RELEASE;
+			}
+			// update volume and frequency as appropriate
+			this->_waveform->setVolume(this->getVolume(elapsed));
+			break;
+		}
+		case AUDIO_STATE_RELEASE: {
+			word elapsed = millis() - this->_startTime;
+			byte newVolume = this->getVolume(elapsed);
+
+			if (isFinished(elapsed)) {
+				// we've reached zero volume, so stop playback
+				this->_waveform->enable(false);
+				debug_log("audio_driver: end (released)\n\r");
+				this->_state = AUDIO_STATE_IDLE;
+			} else {
+				// update volume
+				this->_waveform->setVolume(newVolume);
+			}
+			break;
+		}
+		case AUDIO_STATE_ABORT:

```

Comment:
the abort state is used when something is done that needs to cause playback to abort, such as swapping the waveform on a channel that is playing a note.